File: java/src/solutions/1477_find_two_non_overlapping_sub_arrays_each_with_target_sum.java
import java.util.Arrays;

class Solution {
  public int minSumOfLengths(int[] arr, int target) {
    var n = arr.length;
    var dp = new int[n + 1];
    Arrays.fill(dp, n + 1);
    int result = n + 1, left = 0, count = 0;
    for (var right = 0; right < n; right++) {
      count += arr[right];
      dp[right + 1] = dp[right];

      while (count > target) {
        count -= arr[left];
        left++;
      }

      if (count == target) {
        var len = right - left + 1;
        dp[right + 1] = Math.min(dp[right], len);
        result = Math.min(result, dp[left] + len);
      }
    }

    return result == n + 1 ? -1 : result;
  }
}

File: rust/src/bin/1477_find_two_non_overlapping_sub_arrays_each_with_target_sum.rs
impl Solution {
    pub fn min_sum_of_lengths(arr: Vec<i32>, target: i32) -> i32 {
        let n = arr.len();
        let mut dp = vec![n + 1; n + 1];
        let mut current_count = 0;
        let mut left_idx = 0;
        let mut result = n + 1;
        for right_idx in 0..n {
            current_count += arr[right_idx];
            dp[right_idx + 1] = dp[right_idx];

            while current_count > target {
                current_count -= arr[left_idx];
                left_idx += 1;
            }

            if current_count == target {
                let len = right_idx - left_idx + 1;
                dp[right_idx + 1] = dp[right_idx].min(len);
                result = result.min(len + dp[left_idx]);
            }
        }
        if result == n + 1 { -1 } else { result as i32 }
    }
}

struct Solution {}
